### PR TITLE
Implement weak reference mechanizm for the IndexSpec - [MOD-4733]

### PR DESCRIPTION
**Pull request by GuyAv46**
To improve the handle of an index's lifetime with multi-threading, and to enable a better and uniform way to access an index that may be invalid (already freed), we introduce a few new structures named `IndexSpecManager`, `WeakRef` and `StrongRef`.
1. `IndexSpecManager` wraps an `IndexSpec` and manages both its own refcount and the spec refcount.
2. `WeakRef` wraps a pointer to `IndexSpecManager`, and handles it through its API:
    1. `WeakRef WeakRef_Clone(WeakRef ref);` - returns a new weak reference (increases weak refcount)
    2. `StrongRef WeakRef_Promote(WeakRef w_ref);` - attempts to promote a weak ref into a strong ref. it will return a new strong reference (increase the strong refcount) if the spec that is managed by its manager is still valid, or a strong ref that wraps a NULL pointer if the spec is invalid. It does not consume the weak reference so the strong one should be released when it's no longer needed.
    3. `void WeakRef_Release(WeakRef w_ref);` releases the weak reference
3. `StrongRef` wraps a pointer to `IndexSpecManager` as well, and handles it through its API:
    1.  `StrongRef StrongRef_Clone(StrongRef ref);` - returns a new strong reference (increases strong refcount). Assuming the strong reference was valid, this function does not fail
    2. `WeakRef StrongRef_Demote(StrongRef s_ref);` - returns a new weak reference (increases weak refcount). It does not consume the strong reference.
    3. `void StrongRef_Release(StrongRef s_ref);` - releases the strong reference
    4. `struct IndexSpec *StrongRef_Get(StrongRef s_ref);` - the big difference - returns the underlying index spec.

A common use case for the new weak and strong references is when a background task needs a reference to the spec and it may no longer be valid.
```
void DoSomeStuff(BGObj *obj) {
  StrongRef cur_run = WeakRef_Promote(obj->spec_ref);
  IndexSpec *spec = StrongRef_Get(cur_run);
  if (spec) {
    .
    do cool stuff in the background
    .
    StrongRef_Release(cur_run); // Strong ref is valid. we have to release it
  } else {
    clear_future_work_and_free(obj); // freeing function should also release the object's owned weak reference
  }
}
```

The specs dictionary holds and returns an `IndexSpecManager` pointer instead of a pointer to `IndexSpec`, and it should be handled as borrowing the global strong reference to the spec.
Any long-term hold of the spec should be replaced with a weak reference, and the weak reference should be promoted when needed and demoted until the next use. #3336 will modify the fork GC to own a weak reference instead of a strong one for this purpose. This PR modifies the BG indexing scanner and the search query BG context to hold a weak reference.

Future work could be done to generalize the references to handle other structs by passing upon construction all the needed callbacks.

### CODE REVIEW

File: src/aggregate/aggregate_exec.c [hunk 1]
```diff
@@ -351,14 +353,31 @@ static RedisModuleCtx *blockedClientReqCtx_getRedisctx(const blockedClientReqCtx
...
+    // Notify the client that the query was aborted
+    QueryError status = {0};
+    QueryError_SetError(&status, QUERY_ENOINDEX, "The index was dropped before the query could be executed");
+    QueryError_ReplyAndClear(redis_ctx, &status);
+    blockedClientReqCtx_destroy(BCRctx);
+    return;
+  }
+
+  // lock spec
   RedisSearchCtx_LockSpecRead(BCRctx->req->sctx);
```
Comment (MeirShpilraien) [inline src/aggregate/aggregate_exec.c:381]: Maybe we can create the search ctx here after we know the spec is valid?

Comment (GuyAv46) [reply src/aggregate/aggregate_exec.c:381]: It's created (allocated) before going to the BG. We update its `RedisModuleCtx` first to be linked to the blocked client so we can reply to him if the spec is invalid

File: src/fork_gc.c [hunk 2]
```diff
@@ -71,8 +71,8 @@ static void FGC_unlock(ForkGC *gc, RedisModuleCtx *ctx) {
 static RedisSearchCtx *FGC_getSctx(ForkGC *gc, RedisModuleCtx *ctx) {
   RedisSearchCtx *sctx = NULL;
   if (gc->type == FGC_TYPE_NOKEYSPACE) {
-    sctx = rm_malloc(sizeof(*sctx));
-    *sctx = SEARCH_CTX_STATIC(ctx, gc->sp);
+    // TEMPORARY
+    sctx = NewSearchCtxFromSpec(ctx, gc->sp);
```
Comment (MeirShpilraien) [inline src/fork_gc.c:75]: So this is still not a safe access right? Will be handled on another PR?

Comment (GuyAv46) [reply src/fork_gc.c:75]: Yes. #3336

File: src/module.c [hunk 3]
```diff
@@ -421,8 +415,10 @@ int DropIndexCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
     return RedisModule_WrongArity(ctx);
   }
 
-  IndexSpec *sp = IndexSpec_Load(ctx, RedisModule_StringPtrLen(argv[1], NULL), 0);
-  if (sp == NULL) {
+  const char* spec_name = RedisModule_StringPtrLen(argv[1], NULL);
+  StrongRef global_ref = IndexSpec_LoadUnsafe(ctx, spec_name, 0);
```
Comment (MeirShpilraien) [inline src/module.c:419]: What the name global stands for? Why not spec_ref or spec_strong_ref?

File: src/module.c [hunk 4]
```diff
@@ -519,13 +528,12 @@ int SynUpdateCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
   SynonymMap_UpdateRedisStr(sp->smap, argv + offset, argc - offset, id);
 
   if (initialScan) {
-    IndexSpec_ScanAndReindex(ctx, sp);
+    IndexSpec_ScanAndReindex(ctx, ref);
```
Comment (MeirShpilraien) [inline src/module.c:531]: Shouldn't it get a weak ref?

File: src/module.c (hunk 3 above)
Comment (GuyAv46) [reply src/module.c:419]: because it is a borrow of the global strong ref. can change it

File: src/module.c (hunk 4 above)
Comment (GuyAv46) [reply src/module.c:531]: internally in the creation of the scanner (and only if one is created) it takes a weak ref from this one

File: src/module.c [hunk 5]
```diff
@@ -621,7 +628,7 @@ static int AlterIndexInternalCommand(RedisModuleCtx *ctx, RedisModuleString **ar
       return RedisModule_ReplyWithSimpleString(ctx, "OK");
     }
   }
-  IndexSpec_AddFields(sp, ctx, &ac, initialScan, &status);
+  IndexSpec_AddFields(ref, sp, ctx, &ac, initialScan, &status);
```
Comment (MeirShpilraien) [inline src/module.c:631]: Why does it need to get the ref and not the spec?

Comment (GuyAv46) [reply src/module.c:631]: because this function might start a BG scanning. we can alternatively search a second time in the dictionary

File: src/redisearch_api.c [hunk 6]
```diff
@@ -27,12 +28,13 @@ int RediSearch_GetCApiVersion() {
   return REDISEARCH_CAPI_VERSION;
 }
 
-IndexSpec* RediSearch_CreateIndex(const char* name, const RSIndexOptions* options) {
+RefManager* RediSearch_CreateIndex(const char* name, const RSIndexOptions* options) {
```
Comment (MeirShpilraien) [inline src/redisearch_api.c:31]: Maybe add a general comment that we return the `RefManager*` so we will not have the break to LLAPI but wherever we get it we know that it is actually a `StrongRef`

File: src/rules.c [hunk 7]
```diff
@@ -118,7 +118,7 @@ SchemaRule *SchemaRule_Create(SchemaRuleArgs *args, IndexSpec *spec, QueryError
     rule->prefixes = array_append(rule->prefixes, p);
   }
 
-  rule->spec = spec;
+  rule->spec = StrongRef_Get(ref);
```
Comment (MeirShpilraien) [inline src/rules.c:121]: Here we should also hold the weak or strong ref, right?

File: src/rules.c [hunk 8]
```diff
@@ -389,13 +389,15 @@ int SchemaRule_RdbLoad(IndexSpec *sp, RedisModuleIO *rdb, int encver) {
...
+  SchemaRule *rule = SchemaRule_Create(&args, ref, &status);
   if (!rule) {
     RedisModule_LogIOError(rdb, "warning", "%s", QueryError_GetError(&status));
     RedisModule_Assert(rule);
   } else {
     rule->score_default = score_default;
     rule->lang_default = lang_default;
+    // No need to validate the reference here, since we are loading it from the RDB
+    IndexSpec *sp = StrongRef_Get(ref);
     sp->rule = rule;
```
Comment (MeirShpilraien) [inline src/rules.c:401]: Ohh right, I see the spec holds the rule and the rule holds the spec. this is a classic weak ref usecase.

File: src/rules.c (hunk 7 above)
Comment (GuyAv46) [reply src/rules.c:121]: I'm not sure. I left a TODO in `rules.h` to verify why the rule needs its spec

File: src/spec.c [hunk 9]
```diff
@@ -1059,12 +1085,12 @@ IndexSpec *IndexSpec_Parse(const char *name, const char **argv, int argc, QueryE
...
 
-  return spec;
+  return ref;
 
 failure:  // on failure free the spec fields array and return an error
   spec->flags &= ~Index_Temporary;
-  IndexSpec_Free(spec);
-  return NULL;
+  IndexSpec_RemoveFromGlobals(ref);
+  return (StrongRef){NULL};
```
Comment (MeirShpilraien) [inline src/spec.c:1093]: Why not: `StrongRef{.rm=NULL}`
And avoid the cast

File: src/spec.c [hunk 10]
```diff
@@ -1333,6 +1353,30 @@ void IndexSpec_Free(IndexSpec *spec) {
...
+    // If uniqueid is 0, it means the index was not initialized
+    // and is being freed now during an error.
+    IndexSpec_ClearAliases(ref);
+  }
+
+  SchemaPrefixes_RemoveSpec(ref);
+
+  // mark the spec as deleted and decrement the ref counts owned by the global dictionaries
+  StrongRef_Invalidate(ref);
+  StrongRef_Release(ref);
```
Comment (MeirShpilraien) [inline src/spec.c:1377]: Maybe we should write a comment here why those are the only places that should be freed and explain that other places holds a weak ref:
1. GC
2. Expire timer
3. Background indexing
4. Cursors?

File: src/spec.c [hunk 11]
```diff
@@ -1341,61 +1385,73 @@ void Indexes_Free(dict *d) {
...
   }
   array_free(specs);
 }
 
+
+//---------------------------------------- atomic updates ---------------------------------------
+
+// atomic update of usage counter
+inline static void IndexSpec_IncreasCounter(IndexSpec *sp) {
+  __atomic_fetch_add(&sp->counter , 1, __ATOMIC_SEQ_CST);
```
Comment (MeirShpilraien) [inline src/spec.c:1408]: Why does the spec still needs counter?

File: src/spec.c [hunk 12]
```diff
@@ -1341,61 +1385,73 @@ void Indexes_Free(dict *d) {
...
     if (!sp) {
-      return NULL;
+      return ref;
     }
   }
 
   // Increament the number of uses.
-  // When we move to multi readers this counter needs to be atomic.
-  ++sp->counter;
+  IndexSpec_IncreasCounter(sp);
```
Comment (MeirShpilraien) [inline src/spec.c:1442]: Why needed?

File: src/spec.c (hunk 11 above)
Comment (GuyAv46) [reply src/spec.c:1408]: It's a statistics counter. Not related to refcounts

File: src/spec.c [hunk 13]
```diff
@@ -1799,25 +1856,23 @@ static IndexesScanner *IndexesScanner_New(IndexSpec *spec) {
...
     global_spec_scanner = NULL;
-  } else if (!scanner->cancelled) {
-    if (scanner->spec && scanner->spec->scanner == scanner) {
-      scanner->spec->scanner = NULL;
-      scanner->spec->scan_in_progress = false;
+  } else {
+    StrongRef tmp = WeakRef_Promote(scanner->spec_ref);
+    IndexSpec *spec = StrongRef_Get(tmp);
+    if (spec) {
+      if (spec->scanner == scanner) {
```
Comment (MeirShpilraien) [inline src/spec.c:1863]: I guess this is not needed anymore right?

Comment (MeirShpilraien) [reply src/spec.c:1863]: We can set the scanner to NULL without this check no?

File: src/spec.c [hunk 14]
```diff
@@ -1799,25 +1856,23 @@ static IndexesScanner *IndexesScanner_New(IndexSpec *spec) {
...
-void IndexesScanner_Cancel(IndexesScanner *scanner, bool still_in_progress) {
-  if (scanner->cancelled) {
-    return;
-  }
-  if (!scanner->global && scanner->spec) {
-    scanner->spec->scan_in_progress = still_in_progress;
-    scanner->spec->scanner = NULL;
-    scanner->spec = NULL;
-  }
+void IndexesScanner_Cancel(IndexesScanner *scanner) {
```
Comment (MeirShpilraien) [inline src/spec.c:1875]: Why would we want to cancel? It will be cancelled because we will not be able to promote  the weak ref no?

File: src/spec.c [hunk 15]
```diff
@@ -1832,6 +1887,9 @@ static void IndexSpec_DoneIndexingCallabck(struct RSAddDocumentCtx *docCtx, Redi
 int IndexSpec_UpdateDoc(IndexSpec *spec, RedisModuleCtx *ctx, RedisModuleString *key, DocumentType type);
 static void Indexes_ScanProc(RedisModuleCtx *ctx, RedisModuleString *keyname, RedisModuleKey *key,
                              IndexesScanner *scanner) {
+  if (scanner->cancelled) {
```
Comment (MeirShpilraien) [inline src/spec.c:1890]: I guess this is not needed anymore

File: src/spec.c [hunk 16]
```diff
@@ -2140,7 +2203,7 @@ void Indexes_ScanAndReindex() {
   }
 
   RedisModule_Log(NULL, "notice", "Scanning all indexes");
-  IndexesScanner *scanner = IndexesScanner_New(NULL);
+  IndexesScanner *scanner = IndexesScanner_New((StrongRef){0}, true);
```
Comment (MeirShpilraien) [inline src/spec.c:2206]:
```suggestion
  IndexesScanner *scanner = IndexesScanner_New((StrongRef){.rm = NULL}, true);
```

File: src/spec.c (hunk 14 above)
Comment (GuyAv46) [reply src/spec.c:1875]: It might get canceled when a new scanner replaces a running one, without invalidating the spec. we can "trick" the scanner and replace its reference with an invalid one instead

**Review by MeirShpilraien — COMMENTED** [review]
Big :+1: 
Some small comments/questions.

File: src/rules.c (hunk 7 above)
Comment (MeirShpilraien) [reply src/rules.c:121]: I believe its a weak ref, see my comment bellow.

File: src/spec.c (hunk 11 above)
Comment (MeirShpilraien) [reply src/spec.c:1408]: Yeah, it was needed because the index might be drop and recreated and was wanted to know that on different locations. But I wonder if now its still needed...

File: src/spec.c (hunk 14 above)
Comment (MeirShpilraien) [reply src/spec.c:1875]: Ohh, in case of alter the index? :+1:

File: src/rules.c (hunk 8 above)
Comment (GuyAv46) [reply src/rules.c:401]: check out #3352, it might be a better fix

File: src/rules.c (hunk 7 above)
Comment (GuyAv46) [reply src/rules.c:121]: check out #3352, it might be a better fix

File: src/spec.c (hunk 11 above)
Comment (GuyAv46) [reply src/spec.c:1408]: lets address it in another issue?

File: src/spec.c (hunk 13 above)
Comment (GuyAv46) [reply src/spec.c:1863]: But what if it was replaced and now we free the irrelevant scanner? It should not edit the spec in this case

File: src/spec.c (hunk 14 above)
Comment (GuyAv46) [reply src/spec.c:1875]: Exactly

File: src/spec.c [hunk 17]
```diff
@@ -212,33 +214,52 @@ int isRdbLoading(RedisModuleCtx *ctx) {
...
   if (sp->isTimerSet) {
-    RedisModule_StopTimer(RSDummyContext, sp->timerId, NULL);
+    WeakRef old_timer_ref;
+    if (RedisModule_StopTimer(RSDummyContext, sp->timerId, (void **)&old_timer_ref) == REDISMODULE_OK) {
+      WeakRef_Release(old_timer_ref);
+    }
   }
   sp->timerId = RedisModule_CreateTimer(RSDummyContext, sp->timeout,
-                                        (RedisModuleTimerProc)IndexSpec_TimedOutProc, sp);
+                                        (RedisModuleTimerProc)IndexSpec_TimedOutProc, *(void **)&ref);
```
Comment (DvirDukhan) [inline src/spec.c:253]: isn't `ref` gets out of scope?

File: src/spec.c [hunk 18]
```diff
@@ -1333,6 +1353,30 @@ void IndexSpec_Free(IndexSpec *spec) {
...
 //---------------------------------------------------------------------------------------------
 
+// Assumes this is called from the main thread with no competing threads
+// Also assumes that the spec is existing in the global dictionary, so
+// we use the global reference as our guard and access the spec dierctly.
+// This function consumes the Strong reference it gets
+void IndexSpec_RemoveFromGlobals(StrongRef ref) {
+  IndexSpec *spec = StrongRef_Get(ref);
+
+  // Remove spec from global index list
```
Comment (DvirDukhan) [inline src/spec.c:1363]: isn't it the rm?

File: tests/cpptests/test_cpp_index.cpp [hunk 19]
```diff
@@ -1232,19 +1236,21 @@ TEST_F(IndexTest, testHugeSpec) {
...
+  StrongRef_Release(ref);
   freeSchemaArgs(args);
 
   // test too big a schema
   N = 300;
   fillSchema(args, N);
 
   QueryError_ClearError(&err);
-  s = IndexSpec_Parse("idx", (const char **)&args[0], args.size(), &err);
+  ref = IndexSpec_Parse("idx", (const char **)&args[0], args.size(), &err);
```
Comment (DvirDukhan) [inline tests/cpptests/test_cpp_index.cpp:1252]: release the ref?

File: src/module.c [hunk 20]
```diff
@@ -445,8 +441,21 @@ int DropIndexCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
     keepDocs = 1;
   }
 
-  RedisSearchCtx sctx = SEARCH_CTX_STATIC(ctx, sp);
-  Redis_DropIndex(&sctx, (delDocs || sp->flags & Index_Temporary) && !keepDocs);
+  if((delDocs || sp->flags & Index_Temporary) && !keepDocs) {
+    // We take a strong reference to the index, so it will not be freed
+    // and we can still use it to delete the keys
+    StrongRef oun_ref = StrongRef_Clone(global_ref);
```
Comment (DvirDukhan) [inline src/module.c:447]: do we still need to take it under the assumption of main thread?

File: src/rules.h [hunk 21]
```diff
@@ -66,8 +66,8 @@ typedef struct SchemaRule {
 void SchemaRuleArgs_Free(SchemaRuleArgs *args);
 void LegacySchemaRulesArgs_Free(RedisModuleCtx *ctx);
 
-SchemaRule *SchemaRule_Create(SchemaRuleArgs *args, struct IndexSpec *spec, QueryError *status);
-void SchemaRule_FilterFields(SchemaRule *rule);
+SchemaRule *SchemaRule_Create(SchemaRuleArgs *args, StrongRef ref, QueryError *status);
```
Comment (DvirDukhan) [inline src/rules.h:69]: rename ref

File: src/rules.h [hunk 22]
```diff
@@ -66,8 +66,8 @@ typedef struct SchemaRule {
...
+void SchemaRule_FilterFields(struct IndexSpec *sp);
```
Comment (DvirDukhan) [inline src/rules.h:70]: spec or ref?

File: src/redisearch_api.c (hunk 6 above)
Comment (DvirDukhan) [reply src/redisearch_api.c:38]: Thought: Since `RSIndex` is `RefManager`, consider using `RSIndex` as the type to process in the functions of this file

File: src/util/references.c [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+  __atomic_add_fetch(&rm->weak_refcount, 1, __ATOMIC_RELAXED);
+}
+
+// Returns false if the spec is being freed or marked as invalid,
+// otherwise increases the strong refcount and returns true.
+// Assumes the caller has a weak reference
+static bool RefManager_TryGetStrongReference(RefManager *rm) {
+  uint16_t cur_ref = -1;
+  // Attempt to increase the strong refcount by 1 if it is not 0
+  while (!__atomic_compare_exchange_n(&rm->strong_refcount, &cur_ref, cur_ref + 1, 0, 0, 0)) {
```
Comment (DvirDukhan) [inline src/util/references.c:79]: do while?

File: src/util/references.c [hunk 24]
```diff
@@ -0,0 +1,143 @@
...
+  RefManager_GetWeakReference(ref.rm);
+  return new_ref;
+}
+
+void StrongRef_Invalidate(StrongRef s_ref) {
+  RefManager_InvalidateSpec(s_ref.rm);
+}
+
+void StrongRef_Release(StrongRef s_ref) {
+  RefManager_ReturnReferences(s_ref.rm);
```
Comment (DvirDukhan) [inline src/util/references.c:133]: are you sure that when calling this function one gives up the weak ref?

File: src/util/references.h [hunk 25]
```diff
@@ -0,0 +1,51 @@
...
+ * @brief This file defines a set of reference types that can be used to handle references to IndexSpecs.
+ * The API mimics some of RUST's reference types. It can be generalized to handle any struct as long as it has
+ * a method to get a weak reference and a method to get a strong reference, by passing the appropriate callbacks.
+ */
+
+typedef void(*RefManager_Free)(void *obj);
+typedef struct RefManager RefManager;
+
+// For LLAPI and wrappers only. DO NOT USE directly.
+void *__RefManager_Get_Object(RefManager *rm);
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/util/references.h:23]: Not directly, but most of the functions there needs the spec and gets the manager. See the comment in the head of `redisearch_api.c`

File: src/alias.c [hunk 26]
```diff
@@ -89,39 +90,37 @@ int AliasTable_Del(AliasTable *table, const char *alias, IndexSpec *spec, int op
   return REDISMODULE_OK;
 }
 
-IndexSpec *AliasTable_Get(AliasTable *tbl, const char *alias) {
+StrongRef AliasTable_Get(AliasTable *tbl, const char *alias) {
+  StrongRef ret = {0};
```
[the remark this answers is not in the document]
Comment (GuyAv46) [reply src/alias.c:94]: We can consider adding a reference as a 64-bit type a generic dictionary can hold. Otherwise, we must somehow store the value, and the StrongRef is just a pointer.
if the underlying pointer of a strong ref is null, it means the ref is not valid (in this case, we couldn't find the alias).

File: src/spec.c (hunk 17 above)
Comment (GuyAv46) [reply src/spec.c:253]: It is owned by the timer and will be released in the callback or if the timer is stopped, by the one who stopped it.

File: src/spec.c (hunk 18 above)
Comment (GuyAv46) [reply src/spec.c:1363]: it is, but semantically the comment is true

File: tests/cpptests/test_cpp_index.cpp (hunk 19 above)
Comment (GuyAv46) [reply tests/cpptests/test_cpp_index.cpp:1252]: No need, the `ref` is invalid

File: src/module.c (hunk 20 above)
Comment (GuyAv46) [reply src/module.c:447]: yes, the "remove from globals" will release the global reference and the spec could be freed there, and we still need its doc table. I improved the comments.

File: src/rules.h (hunk 22 above)
Comment (GuyAv46) [reply src/rules.h:70]: This function gets a direct spec

File: src/util/references.c (hunk 23 above)
Comment (GuyAv46) [reply src/util/references.c:79]: OK

File: src/util/references.c (hunk 24 above)
Comment (GuyAv46) [reply src/util/references.c:133]: Yes. every strong reference is also a weak reference, so the manager will never get freed before the object it manages.
Releasing a strong reference means decreasing both refcounts.
Improved the comments

**Review by MeirShpilraien — APPROVED** [review]

**Review by DvirDukhan — APPROVED** [review]
